README.md

```diff
@@ -311,6 +311,25 @@ $ terser example.js -c passes=2 -m --mangle-props regex=/_$/
 ```javascript
 var x={o:3,t:1,calc:function(){return this.t+this.o},i:2};console.log(x.calc());
 ```
+Mangle all private class fields (this should generally be safe, unless you are using quoted properties that start with `#`):
```

Comment:
> unless you are using quoted properties that start with `#`

I don't understand this comment. There shouldn't be any conflict between quoted public properties and a private property with the same name.

Reply:
At DevTools, we observed the following breakage:

```js
const colorMapping = {
  "#000000": "black"
}
```

That's because the keys were hexadecimal representations of colors. The fix was easy: refactor to a `Map`, but we regressed nonetheless. Since `terser` doesn't differentiate between quoted properties and class fields, you can't do the mapping only for class fields.

Reply:
I'm still a little confused, sorry. Is there an internal CL link that shows the usage? My ldap is jridgewell@.

I think what you're aiming to do is mangle private properties, but you want to keep your object properties? Specifically, some of your object properties use quoted `'#000000'` hex keys. You've then enabled `regex: /^#/` to mangle the private properties, but found it also mangled your hex keys. Is that right?

If so, I see two parts that can be fixed in your setup. First, there's no need to set `regex: /^#/` to have Terser mangle your private properties. Just enabling property mangling will turn on private property mangling, regardless of your `regex` setting. That's because private properties are lexically mangle-able without issue (and I think that's why @fabiosantoscode is suggesting it be on by default when enabling regular mangling). Second, you might need `properties: { keep_quoted: true }`  (or `'strict'`) to stop Terser from mangling quoted property (but note there's currently a bug which'll be fixed with https://github.com/terser/terser/pull/1058).

Reply:
No, the CL is open-source: https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/2982146 The regression bug was https://bugs.chromium.org/p/chromium/issues/detail?id=1223298 We enabled private class field mangling in https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/2975285

> Just enabling property mangling

But we can't do that, as we still have plenty of code that refers to the unmangled name in a separate context. E.g. we can **only** be certain that we don't break anything with solely mangling with private class fields. So we can't apply the mangling for everything and we need to somehow select only private class fields. The regex solution does exactly that.

I think we could have done the `keep_quoted` solution, but by now everything is working as expected DevTools-concerned.

Reply:
> So we can't apply the mangling for everything and we need to somehow select only private class fields. The regex solution does exactly that.

Private property mangling doesn't use `regex` when mangling: https://github.com/terser/terser/blob/d3fb5b03c03ed060ac72fe8ef2db6c991af080cb/lib/propmangle.js#L240-L246

You could actually get only private property mangling by setting `regex` to something that never matches. `regex: /$./` will prevent any public properties from being mangled, but because private properties don't care about the regex, they'll still be mangled.

- - -

To get this back on topic for the PR,  I think we can update the description here to demonstrate that private properties will be mangled if `--mangle-props` is used. Though Fabio might prefer we not merge this, since it should be moved form property mangling into regular mangling.

Reply:
Wait, I am getting myself confused. The docs don't mention anything about private property mangling: https://terser.org/docs/cli-usage#cli-mangle-options Does `terser` already do that?

> will prevent any public properties from being mangled, but because private properties don't care about the regex, they'll still be mangled.

Okay, that sounds fine to me. From a user perspective, that is quite weird though. And it isn't mentioned anywhere in the docs. Can we improve `terser` to do this by default so that we don't have to do this hacky solution with the `regex`? (I think we are in agreement on this, but the context of private class mangling wasn't clear to me, as I am not familiar with the codebase)

Reply:
> Does terser already do that?

[Yup](https://github.com/terser/terser/blob/master/CHANGELOG.md#v560)!

> From a user perspective, that is quite weird though. And it isn't mentioned anywhere in the docs. Can we improve terser to do this by default so that we don't have to do this hacky solution with the regex? 

Yes, I think that's the ideal solution.